File: Data Structures and Algorithms/Sorting/Merge Sort/merge_sort.py
import numpy as np 
unsorted = np.random.randint(20, size=10)
print(unsorted)

def mergesort(numlist):
	numlist = list(numlist)
	if len(numlist) > 1:
		l, r = numlist[len(numlist)//2:], numlist[:len(numlist)//2]
		l = mergesort(l) # left section 
		r = mergesort(r) # right section
		i,j,k = 0,0,0 # index for left, right and whole section of the input array

		while i < len(l) and j < len(r):
			if l[i] <= r[j]:
				numlist[k] = l[i]
				i += 1
			else:
				numlist[k] = r[j]
				j += 1
			k += 1
		"""
		# add the remaining values of the left and right section 
		# of the merge sort
		"""
		while i < len(l):
			numlist[k] = l[i]
			i += 1
			k += 1
		while j < len(r):
			numlist[k] = r[j]
			j += 1
			k += 1

	return numlist


		
sorted_list = mergesort(unsorted)
print(sorted_list)

File: Data Structures and Algorithms/Sorting/Merge Sort/MergeSort.java
package sort;

import java.util.Arrays;

public class MergeSort {
	
	private int[] array;
	public void MergesSort(int[] array) {
		this.array = array;
	}
	
	public int[] getArray() {
		return array;
	}
	
	public void setArray(int[] array) {
		this.array = array;
	}
	
	private int[] merge_sort(int[] arr) {
		int array_len = arr.length;
		if(array_len > 1) {
			int[] left = java.util.Arrays.copyOfRange(arr, 0, array_len/2);
			int[] right = java.util.Arrays.copyOfRange(arr, array_len/2, array_len);
			left = merge_sort(left);
			right = merge_sort(right);
			int i = 0,j = 0,k = 0;
			
			while(i < left.length && j < right.length) {
				if(left[i] <= right[j]) {
					arr[k] = left[i];
					i ++;
				}else {
					arr[k] = right[j];
					j ++;
				}
				k += 1;
			}
			while(i < left.length) {
				arr[k] = left[i];
				i ++; k++;			
			}
			while(j < right.length) {
				arr[k] = right[j];
				j++; k++;
			}
		}	
		return arr;		
	}
	
	public int[] sort() {
		return this.merge_sort(this.array);
	}

}
